packages/chain-controller/src/SnapHandlerClient.ts

```diff
@@ -0,0 +1,116 @@
+import { type Sender } from '@metamask/keyring-api';
+import type { JsonRpcRequest } from '@metamask/keyring-api/dist/JsonRpcRequest';
+import type { SnapController } from '@metamask/snaps-controllers';
+import type { SnapId } from '@metamask/snaps-sdk';
+import { HandlerType } from '@metamask/snaps-utils';
+import type { Json } from '@metamask/utils';
+import { v4 as uuid } from 'uuid';
+
+/**
+ * Handler for Snap requests.
+ */
+export type Handler = SnapController['handleRequest'];
+
+/**
+ * Implementation of the `Sender` interface that can be used to send requests
+ * to a snap through a Snap request handler.
+ */
+class SnapHandlerSender implements Sender {
+  #snapId: SnapId;
+
+  #origin: string;
+
+  #handler: Handler;
+
+  #handlerType: HandlerType;
+
+  /**
+   * Create a new instance of `SnapHandlerSender`.
+   *
+   * @param handler - The Snap request handler to send requests to.
+   * @param handlerType - The handler type.
+   * @param snapId - The ID of the snap to use.
+   * @param origin - The sender's origin.
+   */
+  constructor(
+    handler: Handler,
+    handlerType: HandlerType,
+    snapId: SnapId,
+    origin: string,
+  ) {
+    this.#snapId = snapId;
+    this.#origin = origin;
+    this.#handler = handler;
+    this.#handlerType = handlerType;
+  }
+
+  /**
+   * Send a request to the snap and return the response.
+   *
+   * @param request - JSON-RPC request to send to the snap.
+   * @returns A promise that resolves to the response of the request.
+   */
+  async send(request: JsonRpcRequest): Promise<Json> {
+    return this.#handler({
+      snapId: this.#snapId,
+      origin: this.#origin,
+      handler: this.#handlerType,
+      request,
+    }) as Promise<Json>;
```

Comment:
Is there a reason why a type assertion needs to be used? It would be nice to not use them.

Reply:
Well, `this.#handler` will returns a `Promise<unknown>` and I wanted to restrict the type a bit more.

In anyway, given the "dynamic nature" of submitting requests to a Snap through the `handleRequest` handler, having an `unknown` makes sense.

We do have another type assertion here which will be necessary to match our API: https://github.com/MetaMask/core/blob/4ce91aa555a0a688f2f371ae5547738c645d2d13/packages/chain-controller/src/SnapChainProviderClient.ts#L30

So I can probably get rid of this one yes!

Reply:
I did remove the type assertion as we are doing this in the "API client" (see above)